File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/core/BaseFragment.java
package com.przemyslawlusnia.vocabularycreator.core;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import butterknife.Unbinder;
import com.przemyslawlusnia.vocabularycreator.core.utils.ObjectUtils;

public class BaseFragment extends Fragment {

  private static final String TAG = BaseFragment.class.getSimpleName();

  /**
   * There are two types of requests: on demand (click) and on enter screen.
   * <p>
   * On demand: use {@link BaseFragment#networking} to keep info if request has started.
   * If config changes (screen rotation, language change, keyboard change etc...) you will be able to retrieve it
   * with savedInstanceState and restart from onResume() if mentioned value is true.
   * <p>
   * On enter screen: call request from onCreate or onResume (which you prefer for your case)
   */
  protected boolean networking = false;
  protected static final String NETWORKING_STATE = "NETWORKING_STATE";

  protected Unbinder butterknifeUnbinder;
  protected EventListener eventListener;

  public BaseFragment() {
    // not used
  }

  @Override
  public void onAttach(Context context) {
    super.onAttach(context);
    if (context instanceof EventListener) {
      eventListener = (EventListener) context;
    } else {
      Log.e(TAG, "onAttach, context is not instanceof EventListener");
    }
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    if (savedInstanceState != null) {
      networking = savedInstanceState.getBoolean(NETWORKING_STATE);
    }
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putBoolean(NETWORKING_STATE, networking);
  }

  protected void setToolbarTitle(String txt) {
    if (ObjectUtils.isNotNull("setToolbarTitle", TAG, eventListener)) {
      eventListener.setToolbarTitleTxt(txt);
    }
  }

  protected void setToolbarBackground(@ColorRes int colorRes) {
    if (ObjectUtils.isNotNull("setToolbarBackground", TAG, eventListener)) {
      eventListener.setToolbarBackground(colorRes);
    }
  }

  protected void setToolbarShadow(boolean visible) {
    if (ObjectUtils.isNotNull("setToolbarShadow", TAG, eventListener)) {
      eventListener.setToolbarShadow(visible);
    }
  }

  protected void checkListenerClass(Context context, Class<?> expectedContextClass) {
    if (!expectedContextClass.isAssignableFrom(context.getClass())) {
      throw new RuntimeException(context.toString() + " must implement " + expectedContextClass.getSimpleName());
    }
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    if (butterknifeUnbinder != null) {
      butterknifeUnbinder.unbind();
    }
  }

  @Override
  public void onDetach() {
    eventListener = null;
    super.onDetach();
  }

  interface EventListener {
    void setToolbarTitleTxt(String txt);

    void setToolbarBackground(@ColorRes int colorRes);

    void setToolbarShadow(boolean visible);
  }

}

File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/core/BaseFragment.kt
package com.przemyslawlusnia.vocabularycreator.core

import android.content.Context
import android.os.Bundle
import android.support.annotation.ColorRes
import android.support.v4.app.Fragment
import android.util.Log

open class BaseFragment : Fragment() {

    /**
     * There are two types of requests: on demand (click) and on enter screen.
     *
     *
     * On demand: use [BaseFragment.networking] to keep info if request has started.
     * If config changes (screen rotation, language change, keyboard change etc...) you will be able to retrieve it
     * with savedInstanceState and restart from onResume() if mentioned value is true.
     *
     *
     * On enter screen: call request from onCreate or onResume (which you prefer for your case)
     */
    protected var networking = false
    protected var eventListener: EventListener? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is EventListener) {
            eventListener = context
        } else {
            Log.d(TAG, "onAttach, context is not EventListener")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        networking = savedInstanceState?.getBoolean(NETWORKING_STATE) ?: false
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putBoolean(NETWORKING_STATE, networking)
    }

    protected fun setToolbarTitle(txt: String) {
        eventListener?.setToolbarTitleTxt(txt)
    }

    protected fun setToolbarBackground(@ColorRes colorRes: Int) {
        eventListener?.setToolbarBackground(colorRes)
    }

    protected fun setToolbarShadow(visible: Boolean) {
        eventListener?.setToolbarShadow(visible)
    }

    protected fun checkListenerClass(context: Context, expectedContextClass: Class<*>) {
        if (!expectedContextClass.isAssignableFrom(context.javaClass)) {
            throw RuntimeException(context.toString() + " must implement " + expectedContextClass.simpleName)
        }
    }

    override fun onDetach() {
        eventListener = null
        super.onDetach()
    }

    interface EventListener {
        fun setToolbarTitleTxt(txt: String)

        fun setToolbarBackground(@ColorRes colorRes: Int)

        fun setToolbarShadow(visible: Boolean)
    }

    companion object {
        private val TAG = BaseFragment::class.java.simpleName
        protected val NETWORKING_STATE = "NETWORKING_STATE"
    }

}
